File: 07-interfaces/java/Nadador.java
/*
 * Nadador.java - INTERFAZ
 *
 * Una interfaz NO dice lo que algo ES. Dice lo que algo SABE HACER.
 *
 *   Animal   -> ES UN        (herencia)
 *   Nadador  -> SABE NADAR   (interfaz)
 *
 * Fijate que no tiene atributos ni constructor: no guarda estado,
 * solo exige capacidades. Los metodos van sin cuerpo y sin "public":
 * en una interfaz todo es public y abstract automaticamente.
 */
public interface Nadador {

    void nadar();

    double velocidadEnAgua();   // km/h
}

File: 07-interfaces/python/nadador.py
"""
nadador.py - "INTERFAZ"

Python no tiene la palabra `interface`. El equivalente mas cercano es
una clase abstracta que SOLO declara metodos: sin atributos, sin
constructor, sin estado.

  interface Nadador { }   ->   class Nadador(ABC):
  implements Nadador      ->   se agrega como clase base extra

  Animal   -> ES UN        (herencia)
  Nadador  -> SABE NADAR   (interfaz)
"""

from abc import ABC, abstractmethod


class Nadador(ABC):

    @abstractmethod
    def nadar(self):
        """Contrato: quien firme esto, sabe nadar."""

    @abstractmethod
    def velocidad_en_agua(self):
        """km/h"""
